model/messages/consensus.go

```diff
@@ -4,11 +4,132 @@ import (
+	for _, chunk := range internal.Chunks {
+		result.Chunks = append(result.Chunks, *chunk)
+	}
+	return result
+}
+
+// UntrustedBlockPayload is a duplicate of flow.Payload used within
+// untrusted messages. It exists only to provide a memory-safe structure for
+// decoding messages and should be replaced in the future by updating the core
+// flow.Payload type.
+// Deprecated: Please update flow.Payload to use []flow.Guarantee etc., then
+// replace instances of this type with flow.Payload
+type UntrustedBlockPayload struct {
+	Guarantees []flow.CollectionGuarantee
+	Seals      []flow.Seal
+	Receipts   []flow.ExecutionReceiptMeta
+	Results    []UntrustedExecutionResult
+}
+
+// UntrustedBlock is a duplicate of flow.Block used within
+// untrusted messages. It exists only to provide a memory-safe structure for
+// decoding messages and should be replaced in the future by updating the core
+// flow.Block type.
+// Deprecated: Please update flow.Payload to use []flow.Guarantee etc., then
+// replace instances of this type with flow.Block
+type UntrustedBlock struct {
+	Header  flow.Header
+	Payload UntrustedBlockPayload
+}
+
+// ToInternal returns the internal representation of the type.
+func (ub UntrustedBlock) ToInternal() *flow.Block {
+	block := flow.Block{
+		Header:  &ub.Header,
+		Payload: &flow.Payload{},
+	}
+	for _, guarantee := range ub.Payload.Guarantees {
+		guarantee := guarantee
+		block.Payload.Guarantees = append(block.Payload.Guarantees, &guarantee)
+	}
+	for _, seal := range ub.Payload.Seals {
+		seal := seal
+		block.Payload.Seals = append(block.Payload.Seals, &seal)
+	}
+	for _, receipt := range ub.Payload.Receipts {
+		receipt := receipt
+		block.Payload.Receipts = append(block.Payload.Receipts, &receipt)
+	}
+	for _, result := range ub.Payload.Results {
+		result := result
+		block.Payload.Results = append(block.Payload.Results, result.ToInternal())
+	}
+
+	return &block
+}
+
+// UntrustedBlockFromInternal converts the internal flow.Block representation
+// to the representation used in untrusted messages.
+func UntrustedBlockFromInternal(flowBlock *flow.Block) UntrustedBlock {
+	block := UntrustedBlock{
+		Header: *flowBlock.Header,
+	}
+	for _, guarantee := range flowBlock.Payload.Guarantees {
+		block.Payload.Guarantees = append(block.Payload.Guarantees, *guarantee)
+	}
+	for _, seal := range flowBlock.Payload.Seals {
+		block.Payload.Seals = append(block.Payload.Seals, *seal)
+	}
+	for _, receipt := range flowBlock.Payload.Receipts {
+		block.Payload.Receipts = append(block.Payload.Receipts, *receipt)
+	}
+	for _, result := range flowBlock.Payload.Results {
+		block.Payload.Results = append(block.Payload.Results, UntrustedExecutionResultFromInternal(result))
+	}
+	return block
+}
+
+// BlockProposal is part of the consensus protocol and represents the leader
 // of a consensus round pushing a new proposal to the network.
 type BlockProposal struct {
-	Header  *flow.Header
-	Payload *flow.Payload
+	Block UntrustedBlock
```

Comment:
Just for curiosity, would there be a disadvantage to writing
```golang
type BlockProposal struct {
	UntrustedBlock
}
```
?

Reply:
Not really, either seem OK to me.